src/mapping/parser/grammar.pest

```diff
@@ -42,37 +42,15 @@ path_coalesce = !{ "(" ~ (path_segment | quoted_path_segment) ~ ("|" ~ (path_seg
 
 dot_path = ${ ("." ~ (path_segment | quoted_path_segment | path_coalesce))+ }
 
-// Query functions
-//
-// NOTE: Not including delete here yet, but there's no reason we can't slot it
-// in at some point and get `.nested = del(.foo)`, same with `only_fields`.
-query_function = {
-    to_string |
-    to_int |
-    to_float |
-    to_bool |
-    to_timestamp |
-    parse_timestamp |
-    upcase |
-    downcase |
-    uuid_v4 |
-    sha1 |
-    md5 |
-    now
-}
+ident = @{ ASCII_ALPHA_LOWER ~ ( ASCII_ALPHA_LOWER | "_" )* }
+
+function = { ident ~ "(" ~ function_arguments? ~ ")"  }
```

Comment:
Alternatively, as the original issue (https://github.com/timberio/vector/issues/3851) describes, we could just make non-optional arguments positional, and optional ones named, but I'm personally not a fan of that approach, since the named arguments add more clarity to the function call, regardless if the argument is required or not.

cc @binarylogic.

Reply:
There is also the Python approach, which may be more useful when there are more than two possible arguments: any argument may be named, but positional parameters fill in the function parameters in declaration order while named parameters may go in any order (filling in parameters is done at run time by the function, which knows the names). Obviously we want to minimize the run time cost, so determining what goes where has to happen during parsing.

I'd actually tend to agree with your last alternative (that you're not a fan of), though it's not a strong opinion.

Reply:
Yeah, that makes sense. w.r.t. named arguments / positional arguments, that mostly boils down to whatever preference you have on the subject, I guess.

Reply:
Yeah, I like Bruce's suggestion, but it seems more complex. I'm also ok with positional arguments not being named, which is Ruby-esq 😄 👍 

```
method(arg1, arg2, optional1=null, optional2=true)
```

Whatever is easier to maintain is my vote.

Reply:
I like Python's approach. It seems to strike a balance to still allow use of named parameters for required arguments and allows the author to decide when supplying the name of the argument is useful. For example, they may decide to write any of:

* `sha1(.foo, "SHA-384")`
* `sha1(.foo, variant = "SHA-384")`
* `sha1(data = .foo, variant = "SHA-384")`

For documentation and example purposes, we'd probably push people towards `sha1(.foo, variant = "SHA-384")`. I think this would allow for the use of named parameters in situations where the author feels the required arguments are ambiguous in position.

Reply:
> Whatever is easier to maintain is my vote.

I'd argue that whatever provides the best UX would be preferable :smile: Albeit implementation maintenance is also a concern; I'd just weight it lower than end-user UX.

Reply:
> or even `sha1(data = .foo, "SHA-384")` or `sha1(variant = "SHA-384", data = .foo)` or `sha1(variant = "SHA-384", .foo)`

How would this work @jszwedko?

Specifically, how would it know that the second argument became `data` if it's unnamed? Would it check that all other arguments are already named, and that's the only one left and thus assigns it to `data`?

Or, is it still a (named) positional argument, it's just that it skips all named ones for positions, and so `.foo`, being the first unnamed argument, would be the argument that matches position `0`?

Reply:
> Or, is it still a (named) positional argument, it's just that it skips all named ones for positions, and so .foo, being the first unnamed argument, would be the argument that matches position 0?

That's what I was thinking.

This appears to be how Python does it:

> If keyword arguments are present, they are first converted to positional arguments, as follows. First, a list of unfilled slots is created for the formal parameters. If there are N positional arguments, they are placed in the first N slots. Next, for each keyword argument, the identifier is used to determine the corresponding slot (if the identifier is the same as the first formal parameter name, the first slot is used, and so on). If the slot is already filled, a TypeError exception is raised. Otherwise, the value of the argument is placed in the slot, filling it (even if the expression is None, it fills the slot). When all arguments have been processed, the slots that are still unfilled are filled with the corresponding default value from the function definition. (Default values are calculated, once, when the function is defined; thus, a mutable object such as a list or dictionary used as default value will be shared by all calls that don’t specify an argument value for the corresponding slot; this should usually be avoided.) If there are any unfilled slots for which no default value is specified, a TypeError exception is raised. Otherwise, the list of filled slots is used as the argument list for the call.

https://docs.python.org/3/reference/expressions.html#calls

Reply:
Sweet, thanks 🙌 that helps a lot.

Reply:
Technically, Python disallows the last mode, as positional arguments are required to come before any keyword arguments.

I agree that whatever provides the best UX is preferable. However, I think needless complexity tends to degrade the UX, so sometimes stronger rules are better.

Reply:
Jean corrected me, Python requires positional first, so it would allow for:

* `sha1(.foo, "SHA-384")`
* `sha1(.foo, variant = "SHA-384")`
* `sha1(data = .foo, variant = "SHA-384")`
* `sha1(variant = "SHA-384", data = .foo)`

Reply:
(Also Bruce corrected me, I just hadn't seen his comment before I posted mine :smile: )

Reply:
With the latest update, the variants @jszwedko listed now work as expected.

Reply:
I should point out that the variant we discussed before, but is not supported in Python, also works:

* `sha1(variant = "SHA-384", .foo, bla = 1, .bar)`

That is to say, positional arguments are not restricted to the front of the argument list, and they are counted without counting keyword arguments, so in the above example, `.foo` is in positional argument `0`, and `.bar` in `1`.

I do still need to add a check to make sure there's no keyword argument that matches the keyword of a positional argument (or have one be chosen over the other if that's the case)